src/ttl_table.c

```diff
@@ -0,0 +1,142 @@
+    TimeToLiveEntry* entry = (TimeToLiveEntry*)key;
+    array_free(entry->fieldExpirations);
+    rm_free(entry);
+}
+
+static dictType dictTimeToLive = {
+  .hashFunction = hashFunction_DocId,
+  .keyDup = dup_DocId,
+  .valDup = NULL,
+  .keyCompare = compare_DocId,
+  .keyDestructor = destructor_DocId,
+  .valDestructor = destructor_TimeToLiveEntry,
+};
+
+void TimeToLiveTable_Init(TimeToLiveTable *table) {
+    if (!table->hashTable) {
+      table->hashTable = dictCreate(&dictTimeToLive, NULL);
+    }
+}
+
+void TimeToLiveTable_Destroy(TimeToLiveTable *table) {
+    if (table->hashTable) {
+      dictRelease(table->hashTable);
+      table->hashTable = NULL;
+    }
+}
+
+void TimeToLiveTable_Add(TimeToLiveTable *table, t_docId docId, t_expirationTimePoint docExpirationTime, arrayof(FieldExpiration) sortedById) {
+  TimeToLiveEntry* entry = (TimeToLiveEntry*)rm_malloc(sizeof(TimeToLiveEntry));
+  entry->documentExpiration.id = docId;
+  entry->documentExpiration.point = docExpirationTime;
+  entry->fieldExpirations = sortedById;
+  // we don't want the operation to fail so we use dictReplace
+  dictReplace(table->hashTable, (void*)docId, entry);
+}
+
+void TimeToLiveTable_Remove(TimeToLiveTable *table, t_docId docId) {
+  dictDelete(table->hashTable, (void*)docId);
+}
+
+bool TimeToLiveTable_Empty(TimeToLiveTable *table) {
+  return dictSize(table->hashTable) == 0;
+}
+
+static inline bool DidExpire(const t_expirationTimePoint* field, const t_expirationTimePoint* now) {
+  if (!field->tv_sec && !field->tv_nsec) {
+    return false;
+  }
+
+  if (field->tv_sec > now->tv_sec) {
+    return false;
+  } else if (field->tv_sec == now->tv_sec && field->tv_nsec > now->tv_nsec) {
+    return false;
+  } else {
+    return true;
+  }
+}
+
+bool TimeToLiveTable_HasDocExpired(const TimeToLiveTable *table, t_docId docId, const struct timespec* expirationPoint) {
+  dictEntry *entry = dictFind(table->hashTable, (void*)docId);
+  if (!entry) {
+    return false;
+  }
+
+  TimeToLiveEntry* ttlEntry = (TimeToLiveEntry*)dictGetVal(entry);
+  return DidExpire(&ttlEntry->documentExpiration.point, expirationPoint);
+}
+
+static inline bool verifyFieldIndices(const TimeToLiveTable *table, t_docId docId, t_fieldIndex* sortedFieldIndices, size_t fieldCount, enum FieldExpirationPredicate predicate, const struct timespec* expirationPoint) {
+  dictEntry *entry = dictFind(table->hashTable, (void*)docId);
+  if (!entry) {
+    // the document did not have a ttl for itself or its children
+    // if predicate is default then we know at least one field is valid
+    // if predicate is missing then we know the field is indeed missing since the document has no expiration for it
+    return true;
+  }
+
+  TimeToLiveEntry* ttlEntry = (TimeToLiveEntry*)dictGetVal(entry);
+  if (ttlEntry->fieldExpirations == NULL || array_len(ttlEntry->fieldExpirations) == 0) {
+    // the document has no fields with expiration times, there exists at least one valid field
+    return true;
```

Comment:
A document can have an expiration time without having fields with expiration time.
So it would still have an entry in the dict but the fields willl be null/0 sized

Reply:
So why is it uncovered? Can we cover it?

Reply:
We would need an expiration on a document without the document expiration time ever reaching the expiration point.
Will see if I can add such a test.